casper/src/main/scala/coop/rchain/casper/SafetyOracle.scala

```diff
@@ -55,177 +57,29 @@ trait SafetyOracle[F[_]] {
-                    blockDag,
-                    blockHash,
-                    b => b == creatorJustificationOrGenesis
-                  )
-                }
-                .pure[F]
-            case None => StreamT.empty[F, BlockHash].pure[F]
-          }
-
-        def neverEventuallySeeDisagreement(
-            first: Validator,
-            second: Validator
-        ): F[Boolean] =
-          (for {
-            firstLatestBlock <- OptionT(blockDag.latestMessage(first))
-            secondLatestOfFirstLatestHash <- OptionT.fromOption[F](
-                                              firstLatestBlock.justifications
-                                                .find {
-                                                  case Justification(validator, _) =>
-                                                    validator == second
-                                                }
-                                                .map(_.latestBlockHash)
-                                            )
-            secondLatestOfFirstLatest <- OptionT(blockDag.lookup(secondLatestOfFirstLatestHash))
-            potentialDisagreements <- OptionT.liftF(
-                                       filterChildren(secondLatestOfFirstLatest, second)
-                                     )
-            // TODO: Implement forallM on StreamT
-            result <- OptionT.liftF(potentialDisagreements.toList.flatMap(_.forallM {
-                       potentialDisagreement =>
-                         computeCompatibility(blockDag, candidateMetadata, potentialDisagreement)
-                     }))
-          } yield result).fold(false)(identity)
-
-        def computeAgreementGraphEdges: F[List[(Validator, Validator)]] =
-          (for {
-            x <- agreeingValidatorToWeight.keys
-            y <- agreeingValidatorToWeight.keys
-            if x.toString > y.toString // TODO: Order ByteString
-          } yield (x, y)).toList.filterA {
-            case (first: Validator, second: Validator) =>
-              neverEventuallySeeDisagreement(first, second) &&^ neverEventuallySeeDisagreement(
-                second,
-                first
-              )
-          }
-
-        computeAgreementGraphEdges.map { edges =>
-          Clique.findMaximumCliqueByWeight[Validator](edges, agreeingValidatorToWeight)
-        }
-      }
-
-      private def computeCompatibility(
-          blockDag: BlockDagRepresentation[F],
-          candidateMetadata: BlockMetadata,
-          targetBlockHash: BlockHash
-      ): F[Boolean] =
-        isInMainChain(blockDag, candidateMetadata, targetBlockHash)
+      CliqueOracle.normalizedFaultTolerance(
+        blockDag,
+        candidateBlockHash,
+        (h: BlockHash) => blockDag.lookupUnsafe(h).map(_.weightMap),
+        (h: BlockHash) => blockDag.lookupUnsafe(h).map(_.blockNum),
+        (v: Validator) => blockDag.latestMessageHashUnsafe(v),
```

Comment:
Thought here we are reading twice unnecessary - is it that bad? Should `blockDag.lookupUnsafe` result be cached by LMDB in memory?